### PR TITLE
Asterisk fill gaps

### PR DESCRIPTION
**Pull request by roystanmendez**
Changed and fine tune code..

### CODE REVIEW

File: res/res_agi.c [hunk 1]
```diff
@@ -1091,111 +1090,85 @@ static int add_silence(struct ast_channel *chan, struct ast_frame *f, struct ast
...
-		f_ptime=ast_channel_get_s1_ptime(chan);
-                if(f_ptime < 5)
-                        f_ptime=20;
+	/*! ptime for the stream */
+	f_ptime=ast_channel_get_ptime(chan, stream_no);
+	if(f_ptime < 5)
+		f_ptime=20;
+
+	/*! First - Check if it is the 1st packet for the stream and fix the initial delay of streams */
+        if ((ast_channel_get_pkt_count(chan, stream_no) == 0) || (ssrc_change == 1)){
```
Comment (harirakr) [inline res/res_agi.c]: If we re-order this check, we could potentially reduce one function call in most cases. Like,
`if ((ssrc_change || !ast_channel_get_pkt_count(chan, stream_no))`

File: res/res_agi.c [hunk 2]
```diff
@@ -1091,111 +1090,85 @@ static int add_silence(struct ast_channel *chan, struct ast_frame *f, struct ast
         long int ts_diff=0, f_ptime, last_seq=0;
 	int64_t gap_ms=0;
 	struct timeval s_tv = ast_channel_get_rec_start_time(chan);
+	unsigned int themssrc=0;
+	int ssrc_change=0;
+
+        /*! Check SSRC */
+        if ((f->themssrc != 0) && (f->themssrc != ast_channel_get_last_ssrc(chan, stream_no))){
```
Comment (harirakr) [inline res/res_agi.c]: If we are calling the function  `ast_channel_get_last_ssrc` multiple times, we could consider using a local variable instead.

File: res/res_agi.c [hunk 3]
```diff
@@ -1091,111 +1090,85 @@ static int add_silence(struct ast_channel *chan, struct ast_frame *f, struct ast
...
+                ast_debug(1, "STREAM %d ==== SSRC changed (%u) to (%u)\n", stream_no, ast_channel_get_last_ssrc(chan, stream_no), f->themssrc);
+                ast_channel_set_last_ssrc(chan, f->themssrc, stream_no);
+                themssrc = f->themssrc;
+		ssrc_change=1;
+        } else if (f->themssrc == ast_channel_get_last_ssrc(chan, stream_no)) {
+                themssrc = ast_channel_get_last_ssrc(chan, stream_no);
+        }else {
+		//Discard the packet
```
Comment (harirakr) [inline res/res_agi.c]: please add detailed comment on why the packet is discarded so that we can recall later.

File: res/res_agi.c [hunk 4]
```diff
@@ -1091,111 +1090,85 @@ static int add_silence(struct ast_channel *chan, struct ast_frame *f, struct ast
...
+			if ((ssrc_change == 1) && (ast_tvcmp(ast_channel_get_rec_end_ts(chan, stream_no), ast_tv(0, 0)) > 0)){
+				s_tv = ast_channel_get_rec_end_ts(chan, stream_no);	
+			}
+
 			gap_ms = ast_tvdiff_ms(ast_tvnow(), s_tv);
 
-			if ((gap_ms/f_ptime) > 1){
-				ast_log(LOG_NOTICE, "Stream 1 delayed by %ld...\n", gap_ms);
-				ast_log(LOG_WARNING, "STREAM %d -- ts_diff: %ld\t f->ts: %ld\t last_ts: %ld\n", stream_no, gap_ms, f->ts, ast_channel_get_s1_last_ts(chan));
+			if ((gap_ms/f_ptime) > 2){ // Greater than twice the ptime
```
Comment (harirakr) [inline res/res_agi.c]: Should this be `(gap_ms/f_ptime) >= 2`

File: main/channel_internal_api.c [hunk 5]
```diff
@@ -1399,97 +1403,114 @@ int ast_channel_internal_is_finalized(struct ast_channel *chan)
...
-long int ast_channel_get_s2_last_ts(const struct ast_channel *chan)
-{
-        return chan->stream2_last_ts;
-}
-
-void ast_channel_set_s1_last_ts(struct ast_channel *chan, long int ts)
+long int ast_channel_get_last_ts(const struct ast_channel *chan, int stream_no)
 {
-        chan->stream1_last_ts = ts;
+	if (stream_no == 1)
```
Comment (harirakr) [inline main/channel_internal_api.c]: if we use enumeration, we could probably avoid checking the numbers -
```
enum streams {STREAM_1=0, STREAM_2=1};

if (stream_no)
  # return STREAM_2 info
else
  # return STREAM_1 info
```

**Comment by roystanmendez** [comment]
Currently merging the changes, we shall update the code as per recommendation in a later stage

**Pull request merged**